### PR TITLE
EGLWLMockNavigation improvements

### PR DESCRIPTION
**Pull request by michaelrodin**
Hi,

I would like to add improvements to the EGLWLMockNavigation benchmark for using textures (optionally).

Best regards,
Michael Rodin

### CODE REVIEW

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    delete data;
+    return loaded;
+}
+
+// loads an array with image data
+bool TextureLoader::loadArray(void * data, unsigned int width, unsigned int height, unsigned int pixelSizeBits) {
+    if(loaded){
+        // don't load the thexture if a texture was already loaded for this object
+        return false;
+    } else {
```
Comment (eucan) [inline ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp]: you dont need to use "{" if there is only one line of code after it

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/ShaderTexture.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+: ShaderBase(vertexShaderCode, fragmentShaderCode, projectionMatrix)
+{
+    glUseProgram(shaderProgramId);
+    m_uniformModelMatrix = glGetUniformLocation(shaderProgramId, "u_modelMatrix");
+    m_uniformSampler = glGetUniformLocation(shaderProgramId, "s_texture");
+    attributeTexCoord = glGetAttribLocation(shaderProgramId, "a_texCoord");
+}
+
+ShaderTexture::~ShaderTexture()
+{
```
Comment (eucan) [inline ivi-layermanagement-examples/EGLWLMockNavigation/src/ShaderTexture.cpp]: don't you need to delete the shader here ?

Comment (eucan) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/ShaderTexture.cpp]: is it done in base class ?

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp [hunk 3]
```diff
@@ -27,25 +27,28 @@ using std::endl;
...
-, m_index(0, 1, 2)
+, m_index(0, 1, 2, 3)
 , m_pShader(shader)
 {
-	m_vertex[0].x = 0.0;
-	m_vertex[0].y = 0.0;
-	m_vertex[0].z = 0.0;
+    m_vertex[0].x = 0.0;
+    m_vertex[0].y = 0.0;
+    m_vertex[0].z = 0.0;
```
Comment (eucan) [inline ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp]: no whitespace changes pls

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp [hunk 4]
```diff
@@ -93,9 +93,16 @@ Car::Car(vec3f position, vec3f size, vec4f color, ShaderTexture* shader, Texture
...
 {
-    m_pShader->use(&m_position, &m_color);
-
     // draw
+    if (withTexture) {
+        GLuint textureID = texture->getId();
+        ((ShaderTexture *)m_pShader)->use(&m_position, textureID);
+        GLint m_attributeTexCoord = ((ShaderTexture *)m_pShader)->getAttributeTexCoord();
+        glEnableVertexAttribArray(m_attributeTexCoord);
+        glVertexAttribPointer(m_attributeTexCoord, 2, GL_FLOAT, GL_FALSE, 0, m_texCoords);
```
Comment (eucan) [inline ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp]: these gl operations should be done in TextureShader class. It is related to TextureShader

**Comment by eucan** [comment]
I found some issues with scan-build:
/home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp:36:1: warning: identifier ‘nullptr’ is a keyword in C++11 [-Wc++0x-compat]
 , texture(nullptr)
 ^
In file included from /home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/include/Car.h:25:0,
                 from /home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp:20:
/home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/include/TextureLoader.h:23:19: warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
     bool loaded = false;
                   ^
In file included from /home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp:20:0:
/home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/include/Car.h:50:24: warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
     bool withTexture = false;
                        ^
/home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp: In constructor ‘Car::Car(vec3f, vec3f, vec4f, ShaderBase*)’:
/home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp:36:11: error: ‘nullptr’ was not declared in this scope
 , texture(nullptr)
           ^

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+    data = new unsigned char [imageSize];
+    fread(data,1,imageSize,file);
+    fclose(file);
+
+    if(pixelSizeBits == 32){
+        // BMP files with an alpha channel use ABGR format for storing pixels
+        // OpenGLES does not have an ABGR format specifier
+        // therefore change pixel format from ABGR to RGBA
+        for(long long i = 0; i < width*height; i++) {
```
[the remark this answers is not in the document]
Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp]: **width** and **height** are **int**s, so they have the size of 4 bytes. The result of multiplication of two 4 byte variables has the size of 8 bytes. Therefore it's appropriate to choose a long long as the index variable.

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    } else if (pixelSizeBits == 24) {
+        for(long long i = 0; i < width*height; i++) {
+            unsigned char b = data[i*pixelSizeBytes];
+            unsigned char g = data[i*pixelSizeBytes + 1];
+            unsigned char r = data[i*pixelSizeBytes + 2];
+            data[i*pixelSizeBytes]     = r;
+            data[i*pixelSizeBytes + 1] = g;
+            data[i*pixelSizeBytes + 2] = b;
+        }
```
[the remark this answers is not in the document]
Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp]: only 3 out of 8 lines are the same

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    unsigned char header[headerSize];
+    unsigned int dataPos;     // Position in the file where the actual data begins
+    GLsizei width, height;
+    uint16_t pixelSizeBits;
+    uint8_t pixelSizeBytes;
+    // imageSize == width*height*pixelSize, because each pixel consists of pixelSize bytes (red, green and blue)
+    unsigned int imageSize;
+    unsigned char * data;
+    FILE * file = fopen(imagePath,"rb");
+    if(!file){
```
[the remark this answers is not in the document]
Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/TextureLoader.cpp]: Fixed, thanks!

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/ShaderTexture.cpp (hunk 2 above)
Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/ShaderTexture.cpp]: Yes, it is done in ShaderBase destructor, where ShaderBase::destroyShader() is called.

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp (hunk 3 above)
Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp]: I've changed the indentation back to tabs, thank you!

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp (hunk 4 above)
Comment (eucan) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp]: Why do we need to do outside of the class then ? I don't understand your answer

Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/Car.cpp]: Ok, now the classes Car, House and Street use the ShaderTexture class for gl texture operations. Should I move the same gl operations glEnableVertexAttribArray glVertexAttribPointer, which are used for vertices out of those classes into the ShaderBase class as well? I think, this would also make sense, although this was not done this way in the previous benchmark versions.

**Comment by michaelrodin** [comment]
I'm sorry, now it's fixed.

**Comment by eucan** [comment]
I am getting now this warning from static analyzer:
/home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/MockNavi.cpp:172:1: warning: Potential leak of memory pointed to by 'pShaderTexture'
}
^

**Comment by eucan** [comment]
sorry I closed it by mistake. I reopened the PR now

**Comment by michaelrodin** [comment]
> 
> 
> I am getting now this warning from static analyzer:
> /home/eucan/github/oss/wayland-ivi-extension/ivi-layermanagement-examples/EGLWLMockNavigation/src/MockNavi.cpp:172:1: warning: Potential leak of memory pointed to by 'pShaderTexture'
> }
> ^

Now I've added code for saving pShaderTexture and and pShader pointers in the MockNavi object and freeing memory in the destructor. There is no errors in the analyzer regarding the EGLWLMockNavigation any more.

File: ivi-layermanagement-examples/EGLWLMockNavigation/src/MockNavi.cpp [hunk 8]
```diff
@@ -78,27 +118,43 @@ void MockNavi::generateCity()
...
+        }else{
+            obj = new Street(streetPosition, streetSize, streetColor, pShader);
+        }
         m_renderList.push_back(obj);
         m_updateList.push_back(obj);
     }
 
     // generate car
-    vec3f carPosition(1.4 * CITY_GRID_SIZE, 0.001, -0.3);
+    vec3f carPosition(1.4 * CITY_GRID_SIZE, 0.002, -0.3);
```
Comment (kenji-hosokawa) [inline ivi-layermanagement-examples/EGLWLMockNavigation/src/MockNavi.cpp]: The position of the car looks like not center. 1.38 of the X coordinate is better than 1.4. I'm not sure how to calculate it properly.

Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/src/MockNavi.cpp]: The width of the base plate is 3 and the car is in the middle, so if its x position is 1.4 and x size 0.2, it's middle position should be 1.5 and in the middle of the screen. I think, the problem is that the shape of the used texture is not exactly square. 1.39 for the car position with using the texture looks best on my screen :). Now I'm using this position only when a texture could be loaded. And I've fixed car constructor for textures so it now uses the size parameters, which are passed to it.

**Comment by eucan** [comment]
I am getting this log: Image file could not be opened: /usr/share/wayland-ivi-extension/textures/skyscrapers/facade04.bmp

facade04.bmp does not exist

**Comment by eucan** [comment]
It looks good now.
@efriedrich please review

File: ivi-layermanagement-examples/EGLWLMockNavigation/configuration.h.in [hunk 9]
```diff
@@ -0,0 +1 @@
+std::string benchmarkPath = "@CMAKE_INSTALL_PREFIX@/@TEXTURES_DIR@";
```
Comment (efriedrich) [inline ivi-layermanagement-examples/EGLWLMockNavigation/configuration.h.in]: why it is called benchmarkPath, just rename it to texturePath or the like

File: ivi-layermanagement-examples/EGLWLMockNavigation/include/Sky.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+#include "IUpdateable.h"
+#include "vec.h"
+#include "ShaderGradient.h"
+
+class ShaderBase;
+
+class Sky : public IRenderable, public IUpdateable
+{
+public:
+    Sky(vec3f position, vec3f size, vec4f color, ShaderGradient* shader, float maxColorVal = 3.0, float colorChangeSpeed = 0.02);
```
Comment (efriedrich) [inline ivi-layermanagement-examples/EGLWLMockNavigation/include/Sky.h]: seems like color range or color component is defined between 0 and 1.0 but here maxColorVal is 3.0. could you clarify...

Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/include/Sky.h]: The color range should be clamped to the range [0, 1] by opengl but in my experiments it looks like it's not. If i stop incrementing at 1.0 in Sky::update(), I get too much green at the end.

File: ivi-layermanagement-examples/EGLWLMockNavigation/configuration.h.in (hunk 9 above)
Comment (efriedrich) [reply ivi-layermanagement-examples/EGLWLMockNavigation/configuration.h.in]: thanks! looks good

File: ivi-layermanagement-examples/EGLWLMockNavigation/include/Sky.h (hunk 10 above)
Comment (efriedrich) [reply ivi-layermanagement-examples/EGLWLMockNavigation/include/Sky.h]: the open gl will clip it, but the calculations you are doing are not linear, you have a factor 1/2 for green  or 1/5 for blue components maybe this is the reason.

Comment (michaelrodin) [reply ivi-layermanagement-examples/EGLWLMockNavigation/include/Sky.h]: I've modified the Sky class a bit, but unfortunately I could not figure out, why I get a different picture at the end of the sunrise if I set color values, which are over 1.0 to 1.0 at the end of the computation. If the color values would be clipped to 1.0, the picture should not change.
